tests/unit/websocket_test.cc

```diff
@@ -64,3 +81,76 @@ SEASTAR_TEST_CASE(test_websocket_handshake) {
         }
     });
 }
+
+
+
+SEASTAR_TEST_CASE(test_websocket_handler_registration) {
+    return seastar::async([] {
+        loopback_connection_factory factory;
+        loopback_socket_impl lsi(factory);
+
+        auto acceptor = factory.get_server_socket().accept();
+        auto connector = lsi.connect(socket_address(), socket_address());
+        connected_socket sock = connector.get0();
+        auto input = sock.input();
+        auto output = sock.output();
+
+        // Setup server
+        websocket::server ws;
+        ws.register_handler("echo", [] (input_stream<char>& in,
+                        output_stream<char>& out) {
+            return repeat([&in, &out]() {
+                return in.read().then([&out](temporary_buffer<char> f) {
+                    std::cerr << "f.size(): " << f.size() << "\n";
+                    if (f.empty()) {
+                        return make_ready_future<stop_iteration>(stop_iteration::yes);
+                    } else {
+                        return out.write(std::move(f)).then([&out]() {
+                            return out.flush().then([] {
+                                return make_ready_future<stop_iteration>(stop_iteration::no);
+                            });
+                        });
+                    }
+                });
+            });
+        });
+        websocket::connection conn(ws, acceptor.get0().connection);
+        future<> serve = conn.process();
+
+        auto close = defer([&conn, &input, &output, &serve] () noexcept {
+            conn.shutdown();
+            conn.close().get();
+            input.close().get();
+            output.close().get();
+            serve.get();
+         });
+
+        // handshake
+        const std::string request =
+                "GET / HTTP/1.1\r\n"
+                "Upgrade: websocket\r\n"
+                "Connection: Upgrade\r\n"
+                "Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\n"
+                "Sec-WebSocket-Version: 13\r\n"
+                "Sec-WebSocket-Protocol: echo\r\n"
+                "\r\n";
+        output.write(request).get();
+        output.flush().get();
+        input.read().get();
+
+        // Sending and recieving a websocket frame
+        const auto ws_frame = std::string(
+            "\201\204"  // 1000 0001 1000 0100
+            "TEST"      // Masking Key
+            "\0\0\0\0", 10); // Masked Message - TEST
+        const auto rs_frame = std::string(
+            "\201\004" // 1000 0001 0000 0100
+            "TEST", 6);    // Message - TEST
+        output.write(ws_frame).get();
+        output.flush().get();
+
+        auto response = input.read().get();
```

Comment:
That's a mistake - `read()` can read partial data, and here we're interested in receiving all 6 bytes. Use `read_exactly(6)` instead of `read()`, or a manual loop to read until eof. This accidentally works fine now, but once results are returned in multiple smaller batches, it won't work anymore.

Reply:
ref:
```
    /// Reads n bytes from the stream, or fewer if reached the end of stream.
    ///
    /// \returns a future that waits until n bytes are available in the
    /// stream and returns them. If the end of stream is reached before n
    /// bytes were read, fewer than n bytes will be returned - so despite
    /// the method's name, the caller must not assume the returned buffer
    /// will always contain exactly n bytes.
    ///
    /// \throws if an I/O error occurs during the read. As explained above,
    /// prematurely reaching the end of stream is *not* an I/O error.
    future<temporary_buffer<CharType>> read_exactly(size_t n) noexcept;
```